### PR TITLE
NAS-105499 / 12.0 / API Keys

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/api_key.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        except ValueError:
+            return None
+
+        try:
+            db_key = await self.middleware.call("datastore.query", "account.api_key", [("id", "=", key_id)],
+                                                {"get": True})
+        except MatchNotFound:
+            return None
+
+        if not pbkdf2_sha256.verify(key, db_key["key"]):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/api_key.py]: Is this lightweight enough to not block the event loop?

File: src/middlewared/middlewared/restful.py [hunk 2]
```diff
@@ -18,19 +18,28 @@
...
+            raise web.HTTPUnauthorized()
+
+        try:
+            if not await middleware.call('auth.check_user', username, password):
+                raise web.HTTPUnauthorized()
+        except web.HTTPUnauthorized:
+            raise
+        except Exception:
+            raise web.HTTPUnauthorized()
+    elif auth.startswith('API-Key '):
```
Comment (william-gr) [inline src/middlewared/middlewared/restful.py]: This is not a valid authentication schema recognized by IANA, is it?

File: src/middlewared/middlewared/plugins/api_key.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        return db_key
+
+    async def _validate(self, schema_name, data, id=None):
+        verrors = ValidationErrors()
+
+        await self._ensure_unique(verrors, schema_name, "name", data["name"], id)
+
+        if verrors:
+            raise verrors
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/api_key.py]: We jave verrors.check() for shorter, JFYI

File: src/middlewared/middlewared/plugins/api_key.py (hunk 1 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/api_key.py]:
```
[desktop] ~> python3 -m timeit -s 'from passlib.hash import pbkdf2_sha256; s = "*" * 4096; h = pbkdf2_sha256.encrypt("*" * 64)' 'pbkdf2_sha256.verify(s, h)'   
100 loops, best of 3: 15.7 msec per loop
```

File: src/middlewared/middlewared/restful.py (hunk 2 above)
Comment (themylogin) [reply src/middlewared/middlewared/restful.py]: Is `Bearer` ok?

**Pull request merged**